src/shres.c

```diff
@@ -0,0 +1,232 @@
+
+#include <stdio.h>
+#include <string.h>
+
+#include "mega65/shres.h"
+#include "mega65/memory.h"
+
+/// Magic string identifying the SYSPART shared resource area.
+static const unsigned char magic_string[]
+    = { 0x4D, 0x45, 0x47, 0x41, '6', '5',                       // "MEGA65"
+          0x53, 0x48, 0x41, 0x52, 0x45, 0x44,                   // "SHARED"
+          0x52, 0x45, 0x53, 0x4F, 0x55, 0x52, 0x43, 0x45, 0x53, // "RESOURCES"
+          0x00 };
+
+/**
+ * @brief Waits until the SD card is no longer busy.
+ */
+void sdcard_busy_wait(void)
+{
+    while (PEEK(0xD680) & 0x03) {
+        continue;
+    }
+}
+
+/**
+ * @brief Executes the shared resource trap for a given sector number.
+ *
+ * @param arg The sector number to read (as a 32-bit address).
+ * @return 0 on success, 1 if the SD card is busy or the trap fails.
+ */
+char do_shres_trap(unsigned long arg)
+{
+    if (PEEK(0xD680) & 0x03) {
+        return 1;
+    }
+
+    shres_regs[0] = (arg >> 0) & 0xff;
+    shres_regs[1] = (arg >> 8) & 0xff;
+    shres_regs[2] = (arg >> 16) & 0xff;
+    shres_regs[3] = (arg >> 24) & 0xff;
+    shres_trap();
+
+    // Success is indicated by bit 0 = 1 in shres_regs[4]
+    return (shres_regs[4] & 0x01) ^ 0x01;
+}
+
+/**
+ * @brief Opens a shared resource file by name and flags.
+ *
+ * @param resource_name The null-terminated name of the resource to open.
+ * @param required_flags Bitmask of required flags the resource must have.
+ * @param file_handle Pointer to the shared_resource struct to populate.
+ * @return 0 on success, 1 if the resource was not found or an error occurred.
+ */
+char shopen(char* resource_name, unsigned long required_flags,
+    struct shared_resource* file_handle)
+{
+    unsigned int d = shdopen();
+    if (d == 0xffff) {
+        return 1;
+    }
+
+    while (!shdread(required_flags, &d, file_handle)) {
+        if (!strcmp(resource_name, file_handle->name)) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+/**
+ * @brief Reads bytes from an open shared resource.
+ *
+ * @param ptr Pointer to the output buffer to fill.
+ * @param count Number of bytes to read.
+ * @param f Pointer to an open shared_resource structure.
+ * @return Number of bytes actually read.
+ */
+unsigned int shread(
+    unsigned char* ptr, unsigned int count, struct shared_resource* f)
+{
+    unsigned int read_bytes = 0;
+
+    if (!f || f->position >= f->length) {
+        return 0;
+    }
+
+    if (count > (f->length - f->position)) {
+        count = (f->length - f->position);
+    }
+
+    while (count > 0) {
+        unsigned int bytes = 512 - (f->position & 511);
+        if (bytes > count) {
+            bytes = count;
+        }
+
+        // Load the appropriate sector
+        if (do_shres_trap(f->first_sector + (f->position >> 9))) {
+            break;
+        }
+	// And wait for the SD card to actually read it.
+	sdcard_busy_wait();
+
+	// Select SD card buffer, not FDC buffer
+	POKE(0xD689L,PEEK(0xD689L)|0x80);
```

Comment:
Should this be `0xD689` without `L`? The `POKE/PEEK` macros only handle 16-bit addresses as opposed to `lpeek/lpoke`